### PR TITLE
HTTP Generic Connector supports URL parameter injection

### PR DESCRIPTION
**Pull request by BradleyBoutcher**
## What does this PR do?
This PR provides support for `query params` in the config for the generic HTTP connector. This will allow the building of a query string, with any necessary credentials injected into it.

- modified 'renderHeader' function to allow rendering of multiple input fields of the same type
- Moved the validation and saving of template strings to its own function to reduce code reuse
- Added support for passing query parameters through a secretless configuration file, and appending them to the url in the http request

To manually test this, add the `queryParam` section to the `config` subsection of a `secretless.yml` file. Add a dummy key and value (ex. foo: "bar"), and then use secretless as a proxy for an http request. The key and value should be appended to your request, and your response will reflect this.

### What ticket does this PR close?
Connected to #1290 

#### Change log
- [X] The CHANGELOG has been updated, or
- [ ] This PR does not include user-facing changes and doesn't require a CHANGELOG update

#### Test coverage
- [X] This PR includes new unit and integration tests to go with the code changes, or
  - Added unit tests for parsing of query params

- [ ] The changes in this PR do not require tests

#### Documentation
- [ ] This PR does not require updating any documentation, or
- [X] Docs (e.g. `README`s) were updated in this PR, and/or there is a follow-on issue to update docs

#### (For releases only) Manual tests
- [ ] Manually tested [Keychain provider](https://github.com/cyberark/secretless-broker/tree/master/test/providers/keychain)
- [ ] Manually run [Kubernetes-Conjur demo](https://github.com/conjurdemos/kubernetes-conjur-demo) with a local Secretless Broker image build of your branch

### CODE REVIEW

File: internal/plugin/connectors/http/generic/config_test.go [hunk 1]
```diff
@@ -59,6 +64,32 @@ func Test_newConfig(t *testing.T) {
...
+		if err != nil {
+			return
+		}
+
+		// A simple test case
+		assert.Equal(t, "Foo Bar", params["NameWithSpaces"])
+
+		// Assert against value calculated with independent base64 encoder.
+		assert.Equal(t, "c29tZUtleQ==", params["Key"])
+	})
```
Comment (sgnn7) [inline internal/plugin/connectors/http/generic/config_test.go:92]: I'd like to see a couple of more tests here:
- One with special/utf chars so that we know encode is working properly
- One with existing url params so that we can be sure that appending works

File: internal/plugin/connectors/http/generic/config.go [hunk 2]
```diff
@@ -34,34 +35,35 @@ func (c *config) validate(credsByID connector.CredentialValuesByID) error {
 	return nil
 }
 
-// renderedHeaders returns the config's header templates filled in with the
+// Render returns the config's `config` section templates filled in with the
 // given credentialValues.
-func (c *config) renderedHeaders(
+func render(
```
Comment (sgnn7) [inline internal/plugin/connectors/http/generic/config.go:40]: Hmm... why was this changed from being a method on config object?

Comment (sgnn7) [reply internal/plugin/connectors/http/generic/config.go:40]: I think we need a better name than just `render` - maybe we should add one more word to describe what it's actually rendering.

Comment (BradleyBoutcher) [reply internal/plugin/connectors/http/generic/config.go:40]: - Before, the only thing it needed from `config` was access to the Header object. Since we now pass those objects in as a parameter, it seemed unnecessary to keep it as an object specific method. I don't mind either way.
- Agreed on the naming, I can make it more specific

File: internal/plugin/connectors/http/generic/README.md [hunk 3]
```diff
@@ -67,11 +67,16 @@ services:
       password:
         from: conjur
         get: somepassword
+      location:
```
Comment (JakeQuilty) [inline internal/plugin/connectors/http/generic/README.md:70]: Should this be "address" since line 79 uses `.address` or vice versa?

File: internal/plugin/connectors/http/generic/README.md [hunk 4]
```diff
@@ -67,11 +67,16 @@ services:
...
+        from: conjur
+        get: address
     config:
       credentialValidations:
         username: '[^:]+'    # username cannot contain a colon
       headers:
         Authorization: "Basic {{ printf \"%s:%s\" .username .password | base64 }}"
+      queryParams:
+        location: "{{ .address }}"
```
Comment (JakeQuilty) [inline internal/plugin/connectors/http/generic/README.md:79]: See above comment

Comment (BradleyBoutcher) [reply internal/plugin/connectors/http/generic/README.md:79]: fixed!

**Comment by izgeri** [comment]
@BradleyBoutcher please use the PR template next time, it's really handy to make sure we're adding adequate tests, docs, and change log messages with each PR.

File: internal/plugin/connectors/http/generic/config.go [hunk 5]
```diff
@@ -85,23 +87,45 @@ func newConfig(cfgYAML *ConfigYAML) (*config, error) {
...
+	errs validation.Errors) (map[string]*template.Template, validation.Errors) {
+	parsedTemplates := make(map[string]*template.Template)
+	// Validate and save template strings
+	for tmplName, tmplStr := range templates {
+		tmpl := newHTTPTemplate(tmplName)
 		// Ignore pointer to receiver returned by Parse(): it's just "tmpl".
 		_, err := tmpl.Parse(tmplStr)
 		if err != nil {
-			errs[header] = fmt.Errorf("invalid header template: %q", err)
+			errs[tmplName] = fmt.Errorf("invalid query param template: %q", err)
```
Comment (sgnn7) [inline internal/plugin/connectors/http/generic/config.go:109]: This function is run for both headers and query params so this error shouldn't be used for both.

File: internal/plugin/connectors/http/generic/config.go [hunk 6]
```diff
@@ -85,23 +87,45 @@ func newConfig(cfgYAML *ConfigYAML) (*config, error) {
...
+
+	if err := errs.Filter(); err != nil {
+		return nil, err
+	}
+
+	return cfg, nil
+}
+
+func validateAndSaveTemplateString(templates map[string]string,
+	errs validation.Errors) (map[string]*template.Template, validation.Errors) {
```
Comment (sgnn7) [inline internal/plugin/connectors/http/generic/config.go:101]: This is an odd indentation style. Maybe something like this for readability would be better:
```golang
func foo(
    blah int,
    baz int,
)(blahblah int) {

    logicstuff()
}
```

File: internal/plugin/connectors/http/generic/config.go [hunk 7]
```diff
@@ -85,23 +87,45 @@ func newConfig(cfgYAML *ConfigYAML) (*config, error) {
...
+	cfg.QueryParams, errs = validateAndSaveTemplateString(cfgYAML.QueryParams, errs)
+
+	if err := errs.Filter(); err != nil {
+		return nil, err
+	}
+
+	return cfg, nil
+}
+
+func validateAndSaveTemplateString(templates map[string]string,
```
Comment (sgnn7) [inline internal/plugin/connectors/http/generic/config.go:100]: I think something shorter may be needed. How about `unmarshalYAMLParams`? Something else?

File: internal/plugin/connectors/http/generic/config.go [hunk 8]
```diff
@@ -85,23 +87,45 @@ func newConfig(cfgYAML *ConfigYAML) (*config, error) {
...
 
-	if err := errs.Filter(); err != nil {
-		return nil, err
+func appendQueryParams(URL url.URL, params map[string]string) string {
+	var query url.Values
+	if len(URL.RawQuery) > 0 {
+		query = URL.Query()
+	} else {
+		query = url.Values{}
+	}
```
Comment (sgnn7) [inline internal/plugin/connectors/http/generic/config.go:123]: Can this be flattened somehow? Eg:
```golang
query := url.Values{}
if len(URL.RawQuery) > 0  {
    query = URL.Query()
}
```

File: internal/plugin/connectors/http/generic/config_test.go [hunk 9]
```diff
@@ -59,6 +65,32 @@ func Test_newConfig(t *testing.T) {
...
```
Comment (sgnn7) [inline internal/plugin/connectors/http/generic/config_test.go:93]: I still don't see a test here that includes special/utf chars. The one that uses `==` isn't adequate enough. See if you can add a test using chars like [these](https://github.com/cyberark/conjur-api-python3/blob/master/test/test_config/policies/variables.yml#L6).

File: internal/plugin/connectors/http/generic/config.go (hunk 5 above)
Comment (BradleyBoutcher) [reply internal/plugin/connectors/http/generic/config.go:109]: Goooood catch

File: internal/plugin/connectors/http/generic/config.go (hunk 6 above)
Comment (BradleyBoutcher) [reply internal/plugin/connectors/http/generic/config.go:101]: Totally agree, fixed it up 👍

File: internal/plugin/connectors/http/generic/config.go (hunk 7 above)
Comment (BradleyBoutcher) [reply internal/plugin/connectors/http/generic/config.go:100]: I'm not sure that's accurate, as we've already performed our unmarshalling at this point. Here we're converting the strings to templates, so I think something like "stringToTemplate" would be best.

File: internal/plugin/connectors/http/generic/README.md [hunk 10]
```diff
@@ -103,6 +108,41 @@ powerful transformation features.  You can use `printf` for formatting and
...
 above for detailed information on these and other features.
 
+### `queryParams`
+
+Like `headers`, this is another key section. The `queryParams` section
+is used to generate a query string, which is appended to your existing URL
+without replacing any existing query parameters.
+
+The _keys_ of the queryParams are defined by the yaml keys.  In the examples
+above, these query parameter key is `location`.
```
Comment (izgeri) [inline internal/plugin/connectors/http/generic/README.md:118]:
```suggestion
above, the query parameter key is `location`.
```

File: internal/plugin/connectors/http/generic/README.md [hunk 11]
```diff
@@ -103,6 +108,41 @@ powerful transformation features.  You can use `printf` for formatting and
...
+```
+
+You can refer to your credentials in this template using the credential name
+preceded by a `.` (eg, `.address` will refer to the credential
+`address`). At runtime, Secretless will replace these
+credential references with your real credentials.
+
+As you can see in the Basic auth example, the `text/template` package has
+powerful transformation features.  You can use `printf` for formatting and
+compose functions using pipes `|`.  See the text template package docs linked
```
Comment (izgeri) [inline internal/plugin/connectors/http/generic/README.md:143]:
```suggestion
you can compose functions using pipes `|`.  See the text template package docs linked
```
not having the "you can" at the start here led me to try to understand "formatting and compose functions" as a concept
